### PR TITLE
Curve25519 generic keyparsing

### PR DESCRIPTION
**Pull request by bigbrett**
Adds API for curve25519 `der` import/export that can handle both private and public keys in one step. I realize that typically curve25519 keys are distributed as public/private only, since you can recover the public key from the private key, but this API simplifies usage for some cases in that it doesn't require multiple import attempts if you don't know at runtime what the contents of a `der` key holds. 

This API is particularly useful for wolfHSM, where the server doesn't know what the contents of a `DER` key contain when it loads/stores from NVM, but needs to import the key into a `curve25519_key` struct that then will have the proper public/private key fields set. The current API doesn't provide a straightforward way to do this.

Added doxygen, and will update the manual/docs once this merges

# Testing

- Added test to ensure new API can handle private only/public only keys in the same exact way as the existing API
- Added test to ensure a combined keypair import/export yields the same exported der as public/private only

# Checklist

 - [x] added tests
 - [x] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

**Review by dgarske — APPROVED** [review]
Over to @SparkiDev

File: wolfcrypt/src/asn.c [hunk 1]
```diff
@@ -35661,6 +35661,55 @@ int wc_Curve25519PublicKeyDecode(const byte* input, word32* inOutIdx,
...
+
+/* Decode Curve25519 key from DER format - can handle private only,
+ * public only, or private+public key pairs.
+ * return 0 on success, negative on error */
+int wc_Curve25519KeyDecode(const byte* input, word32* inOutIdx,
+                          curve25519_key* key, word32 inSz)
+{
+    int ret;
+    byte privKey[CURVE25519_KEYSIZE];
+    byte pubKey[CURVE25519_KEYSIZE];
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:35673]: CURVE25519_PUB_KEY_SIZE
(same value but more correct name)

File: wolfcrypt/src/asn.c [hunk 2]
```diff
@@ -35868,6 +35917,63 @@ int wc_Curve25519PublicKeyToDer(curve25519_key* key, byte* output, word32 inLen,
...
 }
+
+/* Export Curve25519 key to DER format - handles private only, public only,
+ * or private+public key pairs based on what's set in the key structure.
+ * Returns length written on success, negative on error */
+int wc_Curve25519KeyToDer(curve25519_key* key, byte* output, word32 inLen, int withAlg)
+{
+    int ret;
+    byte privKey[CURVE25519_KEYSIZE];
+    byte pubKey[CURVE25519_KEYSIZE];
```
Comment (SparkiDev) [inline wolfcrypt/src/asn.c:35928]: CURVE25519_PUB_KEY_SIZE

File: wolfcrypt/test/test.c [hunk 3]
```diff
@@ -35042,6 +35042,163 @@ static wc_test_ret_t curve255519_der_test(void)
...
+        else {
+            ret = WC_TEST_RET_ENC_EC(ret);
+        }
+    }
+    if (ret == 0 && (outputSz != (word32)sizeof(kCurve25519PubDer) ||
+                    XMEMCMP(output, kCurve25519PubDer, outputSz) != 0)) {
+        ret = WC_TEST_RET_ENC_NC;
+    }
+
+    /* Test decode/encode of a key file containing both public and private
```
Comment (SparkiDev) [inline wolfcrypt/test/test.c:35101]: key file -> key data?

**Review by SparkiDev — CHANGES_REQUESTED** [review]

File: doc/dox_comments/header_files/asn_public.h [hunk 4]
```diff
@@ -1557,6 +1557,137 @@ int wc_EccPublicKeyToDer(ecc_key* key, byte* output,
...
+    the DER key contains other issues despite being properly formatted.
+    \return BUFFER_E Returns if the input buffer is too small to contain a
+    valid DER encoded key.
+
+    \param input Pointer to buffer containing DER encoded private key
+    \param inOutIdx Index to start reading input buffer from.  On output,
+    index is set to last position parsed of input buffer.
+    \param key Pointer to curve25519_key structure to store decoded key
+    \param inSz Size of input DER buffer
+*/
```
Comment (SparkiDev) [inline doc/dox_comments/header_files/asn_public.h:1581]: Add \sa comments for see also.

File: doc/dox_comments/header_files/asn_public.h [hunk 5]
```diff
@@ -1557,6 +1557,137 @@ int wc_EccPublicKeyToDer(ecc_key* key, byte* output,
...
+int wc_Curve25519PrivateKeyDecode(const byte* input, word32* inOutIdx,
```
Comment (SparkiDev) [inline doc/dox_comments/header_files/asn_public.h:1582]: Missing examples too.

**Review by SparkiDev — DISMISSED** [review]

File: wolfcrypt/src/asn.c (hunk 1 above)
Comment (bigbrett) [reply wolfcrypt/src/asn.c:35673]: fixed

File: wolfcrypt/src/asn.c (hunk 2 above)
Comment (bigbrett) [reply wolfcrypt/src/asn.c:35928]: fixed

File: wolfcrypt/test/test.c (hunk 3 above)
Comment (bigbrett) [reply wolfcrypt/test/test.c:35101]: fixed

File: doc/dox_comments/header_files/asn_public.h (hunk 4 above)
Comment (bigbrett) [reply doc/dox_comments/header_files/asn_public.h:1596]: done

File: doc/dox_comments/header_files/asn_public.h (hunk 5 above)
Comment (bigbrett) [reply doc/dox_comments/header_files/asn_public.h:1597]: added examples

**Comment by bigbrett** [comment]
@dgarske @SparkiDev addressed all feedback

**Review by dgarske — APPROVED** [review]

**Review by SparkiDev — APPROVED** [review]
Great work by the way.

**Comment by bigbrett** [comment]
Thanks @SparkiDev!

**Pull request merged**